src/dataflow/src/source/postgres.rs

```diff
@@ -409,10 +458,67 @@ impl PostgresSourceReader {
+                                        "alter table detected on {} with id {}",
+                                        source_table.name, source_table.relation_id
+                                    );
+                                    return Err(Fatal(anyhow!(
+                                        "source table {} with oid {} has been altered",
+                                        source_table.name,
+                                        source_table.relation_id
+                                    )));
+                                }
+                                let cols = relation.columns().iter().map(|c| PostgresColumn {
+                                    name: c.name().unwrap().to_string(),
+                                    type_oid: c.type_id(),
+                                    type_mod: c.type_modifier(),
+                                    nullable: bool::default(),
+                                    primary_key: bool::default(),
+                                });
+                                // Relation messages do not include nullability/primary_key data
+                                // so we check the name, type_oid, and type_mod explicitly and error if any of them differ
+                                if !source_table.columns.iter().zip(cols).all(|(src, rel)| {
+                                    src.name == rel.name
+                                        && src.type_oid == rel.type_oid
+                                        && src.type_mod == rel.type_mod
+                                }) {
+                                    error!("alter table error: name {}, oid {}, old_schema {:?}, new_schema {:?}", source_table.name, source_table.relation_id, source_table.columns, relation.columns());
+                                    return Err(Fatal(anyhow!(
+                                        "source table {} with oid {} has been altered",
+                                        source_table.name,
+                                        source_table.relation_id
+                                    )));
+                                }
+                            } else {
+                                // We do not care about this table so ignore the message entirely
+                                continue;
+                            }
+                        }
+                        Origin(_) | Type(_) => {
                             self.metrics.ignored.inc();
                         }
-                        Truncate(_) => return Err(Fatal(anyhow!("source table got truncated"))),
+                        Truncate(truncate) => {
+                            let tables = truncate
+                                .rel_ids()
+                                .iter()
+                                // Filter here makes option handling in map "safe"
+                                .filter(|id| self.source_tables.contains_key(&id))
```

Comment:
Use `filter_map` here to and use `.map(|table|` directly below

```suggestion
                                .filter_map(|id| self.source_tables.get(&id))
```

Reply:
Works for me, though I'm unsure why `filter_map` is preferable to `filter` here?

Reply:
It gets rid of the `.unwrap()` below by filtering out None values and passing on the contents of `Some(_)` values